diff -urpN linux-5.4.142.old/include/linux/if_bridge.h linux-5.4.142/include/linux/if_bridge.h
--- linux-5.4.142.old/include/linux/if_bridge.h	2021-08-18 13:57:05.000000000 +0700
+++ linux-5.4.142/include/linux/if_bridge.h	2021-08-27 14:42:54.071069693 +0700
@@ -47,6 +47,7 @@ struct br_ip_list {
 #define BR_BCAST_FLOOD		BIT(14)
 #define BR_NEIGH_SUPPRESS	BIT(15)
 #define BR_ISOLATED		BIT(16)
+#define BR_BLOCK_BPDU		BIT(17)
 
 #define BR_DEFAULT_AGEING_TIME	(300 * HZ)
 
diff -urpN linux-5.4.142.old/include/uapi/linux/if_link.h linux-5.4.142/include/uapi/linux/if_link.h
--- linux-5.4.142.old/include/uapi/linux/if_link.h	2021-08-18 13:57:05.000000000 +0700
+++ linux-5.4.142/include/uapi/linux/if_link.h	2021-08-27 14:42:25.131055050 +0700
@@ -340,6 +340,9 @@ enum {
 	IFLA_BRPORT_NEIGH_SUPPRESS,
 	IFLA_BRPORT_ISOLATED,
 	IFLA_BRPORT_BACKUP_PORT,
+	IFLA_BRPORT_MRP_RING_OPEN,
+	IFLA_BRPORT_MRP_IN_OPEN,
+	IFLA_BRPORT_BLOCK_BPDU,
 	__IFLA_BRPORT_MAX
 };
 #define IFLA_BRPORT_MAX (__IFLA_BRPORT_MAX - 1)
diff -urpN linux-5.4.142.old/net/bridge/br_netlink.c linux-5.4.142/net/bridge/br_netlink.c
--- linux-5.4.142.old/net/bridge/br_netlink.c	2021-08-18 13:57:05.000000000 +0700
+++ linux-5.4.142/net/bridge/br_netlink.c	2021-08-27 14:41:41.771033123 +0700
@@ -152,6 +152,7 @@ static inline size_t br_port_info_size(v
 		+ nla_total_size(sizeof(u8))	/* IFLA_BRPORT_MULTICAST_ROUTER */
 #endif
 		+ nla_total_size(sizeof(u16))	/* IFLA_BRPORT_GROUP_FWD_MASK */
+		+ nla_total_size(1)	/* IFLA_BRPORT_BLOCK_BPDU */
 		+ 0;
 }
 
@@ -214,7 +215,8 @@ static int br_port_fill_attrs(struct sk_
 	    nla_put_u16(skb, IFLA_BRPORT_GROUP_FWD_MASK, p->group_fwd_mask) ||
 	    nla_put_u8(skb, IFLA_BRPORT_NEIGH_SUPPRESS,
 		       !!(p->flags & BR_NEIGH_SUPPRESS)) ||
-	    nla_put_u8(skb, IFLA_BRPORT_ISOLATED, !!(p->flags & BR_ISOLATED)))
+	    nla_put_u8(skb, IFLA_BRPORT_ISOLATED, !!(p->flags & BR_ISOLATED)) ||
+	    nla_put_u8(skb, IFLA_BRPORT_BLOCK_BPDU, !!(p->flags & BR_BLOCK_BPDU)))
 		return -EMSGSIZE;
 
 	timerval = br_timer_value(&p->message_age_timer);
@@ -676,6 +678,7 @@ static const struct nla_policy br_port_p
 	[IFLA_BRPORT_NEIGH_SUPPRESS] = { .type = NLA_U8 },
 	[IFLA_BRPORT_ISOLATED]	= { .type = NLA_U8 },
 	[IFLA_BRPORT_BACKUP_PORT] = { .type = NLA_U32 },
+	[IFLA_BRPORT_BLOCK_BPDU]	= { .type = NLA_U8 },
 };
 
 /* Change the state of the port and notify spanning tree */
@@ -774,6 +777,11 @@ static int br_setport(struct net_bridge_
 	if (err)
 		return err;
 
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_BLOCK_BPDU, BR_BLOCK_BPDU);
+	if (err)
+		return err;
+
+
 	br_vlan_tunnel_old = (p->flags & BR_VLAN_TUNNEL) ? true : false;
 	err = br_set_port_flag(p, tb, IFLA_BRPORT_VLAN_TUNNEL, BR_VLAN_TUNNEL);
 	if (err)
diff -urpN linux-5.4.142.old/net/bridge/br_stp_bpdu.c linux-5.4.142/net/bridge/br_stp_bpdu.c
--- linux-5.4.142.old/net/bridge/br_stp_bpdu.c	2021-08-18 13:57:05.000000000 +0700
+++ linux-5.4.142/net/bridge/br_stp_bpdu.c	2021-08-27 14:40:10.210986795 +0700
@@ -83,6 +83,10 @@ void br_send_config_bpdu(struct net_brid
 	if (p->br->stp_enabled != BR_KERNEL_STP)
 		return;
 
+	if (p->flags & BR_BLOCK_BPDU) {
+		return;
+	}
+
 	buf[0] = 0;
 	buf[1] = 0;
 	buf[2] = 0;
@@ -128,6 +132,10 @@ void br_send_tcn_bpdu(struct net_bridge_
 	if (p->br->stp_enabled != BR_KERNEL_STP)
 		return;
 
+	if (p->flags & BR_BLOCK_BPDU) {
+		return;
+	}
+
 	buf[0] = 0;
 	buf[1] = 0;
 	buf[2] = 0;
@@ -171,6 +179,9 @@ void br_stp_rcv(const struct stp_proto *
 	if (p->state == BR_STATE_DISABLED)
 		goto out;
 
+	if (p->flags & BR_BLOCK_BPDU) {
+		goto out;
+	}
 	if (!ether_addr_equal(eth_hdr(skb)->h_dest, br->group_addr))
 		goto out;
 
diff -urpN linux-5.4.142.old/net/bridge/br_sysfs_if.c linux-5.4.142/net/bridge/br_sysfs_if.c
--- linux-5.4.142.old/net/bridge/br_sysfs_if.c	2021-08-18 13:57:05.000000000 +0700
+++ linux-5.4.142/net/bridge/br_sysfs_if.c	2021-08-27 14:40:10.210986795 +0700
@@ -233,6 +233,7 @@ BRPORT_ATTR_FLAG(multicast_flood, BR_MCA
 BRPORT_ATTR_FLAG(broadcast_flood, BR_BCAST_FLOOD);
 BRPORT_ATTR_FLAG(neigh_suppress, BR_NEIGH_SUPPRESS);
 BRPORT_ATTR_FLAG(isolated, BR_ISOLATED);
+BRPORT_ATTR_FLAG(block_bpdu, BR_BLOCK_BPDU);
 
 #ifdef CONFIG_BRIDGE_IGMP_SNOOPING
 static ssize_t show_multicast_router(struct net_bridge_port *p, char *buf)
@@ -286,6 +287,7 @@ static const struct brport_attribute *br
 	&brport_attr_neigh_suppress,
 	&brport_attr_isolated,
 	&brport_attr_backup_port,
+	&brport_attr_block_bpdu,
 	NULL
 };
 
diff -urpN linux-5.4.142.old/net/core/rtnetlink.c linux-5.4.142/net/core/rtnetlink.c
--- linux-5.4.142.old/net/core/rtnetlink.c	2021-08-18 13:57:05.000000000 +0700
+++ linux-5.4.142/net/core/rtnetlink.c	2021-08-27 14:40:49.707006783 +0700
@@ -4372,7 +4372,9 @@ int ndo_dflt_bridge_getlink(struct sk_bu
 	    brport_nla_put_flag(skb, flags, mask,
 				IFLA_BRPORT_UNICAST_FLOOD, BR_FLOOD) ||
 	    brport_nla_put_flag(skb, flags, mask,
-				IFLA_BRPORT_PROXYARP, BR_PROXYARP)) {
+				IFLA_BRPORT_PROXYARP, BR_PROXYARP) ||
+	    brport_nla_put_flag(skb, flags, mask,
+				IFLA_BRPORT_BLOCK_BPDU, BR_BLOCK_BPDU)) {
 		nla_nest_cancel(skb, protinfo);
 		goto nla_put_failure;
 	}
