diff -urpN linux-5.10.60.old/include/linux/if_bridge.h linux-5.10.60/include/linux/if_bridge.h
--- linux-5.10.60.old/include/linux/if_bridge.h	2021-08-18 13:59:19.000000000 +0700
+++ linux-5.10.60/include/linux/if_bridge.h	2021-08-27 14:29:20.142706927 +0700
@@ -56,6 +56,7 @@ struct br_ip_list {
 #define BR_MRP_AWARE		BIT(17)
 #define BR_MRP_LOST_CONT	BIT(18)
 #define BR_MRP_LOST_IN_CONT	BIT(19)
+#define BR_BLOCK_BPDU		BIT(20)
 
 #define BR_DEFAULT_AGEING_TIME	(300 * HZ)
 
diff -urpN linux-5.10.60.old/include/uapi/linux/if_link.h linux-5.10.60/include/uapi/linux/if_link.h
--- linux-5.10.60.old/include/uapi/linux/if_link.h	2021-08-18 13:59:19.000000000 +0700
+++ linux-5.10.60/include/uapi/linux/if_link.h	2021-08-27 14:28:47.998694373 +0700
@@ -524,6 +524,7 @@ enum {
 	IFLA_BRPORT_BACKUP_PORT,
 	IFLA_BRPORT_MRP_RING_OPEN,
 	IFLA_BRPORT_MRP_IN_OPEN,
+	IFLA_BRPORT_BLOCK_BPDU,
 	__IFLA_BRPORT_MAX
 };
 #define IFLA_BRPORT_MAX (__IFLA_BRPORT_MAX - 1)
diff -urpN linux-5.10.60.old/net/bridge/br_netlink.c linux-5.10.60/net/bridge/br_netlink.c
--- linux-5.10.60.old/net/bridge/br_netlink.c	2021-08-18 13:59:19.000000000 +0700
+++ linux-5.10.60/net/bridge/br_netlink.c	2021-08-27 14:33:09.786796532 +0700
@@ -154,6 +154,7 @@ static inline size_t br_port_info_size(v
 		+ nla_total_size(sizeof(u16))	/* IFLA_BRPORT_GROUP_FWD_MASK */
 		+ nla_total_size(sizeof(u8))	/* IFLA_BRPORT_MRP_RING_OPEN */
 		+ nla_total_size(sizeof(u8))	/* IFLA_BRPORT_MRP_IN_OPEN */
+		+ nla_total_size(1)	/* IFLA_BRPORT_BLOCK_BPDU */
 		+ 0;
 }
 
@@ -220,7 +221,8 @@ static int br_port_fill_attrs(struct sk_
 							  BR_MRP_LOST_CONT)) ||
 	    nla_put_u8(skb, IFLA_BRPORT_MRP_IN_OPEN,
 		       !!(p->flags & BR_MRP_LOST_IN_CONT)) ||
-	    nla_put_u8(skb, IFLA_BRPORT_ISOLATED, !!(p->flags & BR_ISOLATED)))
+	    nla_put_u8(skb, IFLA_BRPORT_ISOLATED, !!(p->flags & BR_ISOLATED)) ||
+	    nla_put_u8(skb, IFLA_BRPORT_BLOCK_BPDU, !!(p->flags & BR_BLOCK_BPDU)))
 		return -EMSGSIZE;
 
 	timerval = br_timer_value(&p->message_age_timer);
@@ -728,6 +730,7 @@ static const struct nla_policy br_port_p
 	[IFLA_BRPORT_NEIGH_SUPPRESS] = { .type = NLA_U8 },
 	[IFLA_BRPORT_ISOLATED]	= { .type = NLA_U8 },
 	[IFLA_BRPORT_BACKUP_PORT] = { .type = NLA_U32 },
+	[IFLA_BRPORT_BLOCK_BPDU]	= { .type = NLA_U8 },
 };
 
 /* Change the state of the port and notify spanning tree */
@@ -826,6 +829,11 @@ static int br_setport(struct net_bridge_
 	if (err)
 		return err;
 
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_BLOCK_BPDU, BR_BLOCK_BPDU);
+	if (err)
+		return err;
+
+
 	br_vlan_tunnel_old = (p->flags & BR_VLAN_TUNNEL) ? true : false;
 	err = br_set_port_flag(p, tb, IFLA_BRPORT_VLAN_TUNNEL, BR_VLAN_TUNNEL);
 	if (err)
diff -urpN linux-5.10.60.old/net/bridge/br_stp_bpdu.c linux-5.10.60/net/bridge/br_stp_bpdu.c
--- linux-5.10.60.old/net/bridge/br_stp_bpdu.c	2021-08-18 13:59:19.000000000 +0700
+++ linux-5.10.60/net/bridge/br_stp_bpdu.c	2021-08-27 14:31:41.462762072 +0700
@@ -83,6 +83,10 @@ void br_send_config_bpdu(struct net_brid
 	if (p->br->stp_enabled != BR_KERNEL_STP)
 		return;
 
+	if (p->flags & BR_BLOCK_BPDU) {
+		return;
+	}
+
 	buf[0] = 0;
 	buf[1] = 0;
 	buf[2] = 0;
@@ -130,6 +134,10 @@ void br_send_tcn_bpdu(struct net_bridge_
 	if (p->br->stp_enabled != BR_KERNEL_STP)
 		return;
 
+	if (p->flags & BR_BLOCK_BPDU) {
+		return;
+	}
+
 	buf[0] = 0;
 	buf[1] = 0;
 	buf[2] = 0;
@@ -175,6 +183,9 @@ void br_stp_rcv(const struct stp_proto *
 	if (p->state == BR_STATE_DISABLED)
 		goto out;
 
+	if (p->flags & BR_BLOCK_BPDU) {
+		goto out;
+	}
 	if (!ether_addr_equal(eth_hdr(skb)->h_dest, br->group_addr))
 		goto out;
 
diff -urpN linux-5.10.60.old/net/bridge/br_sysfs_if.c linux-5.10.60/net/bridge/br_sysfs_if.c
--- linux-5.10.60.old/net/bridge/br_sysfs_if.c	2021-08-18 13:59:19.000000000 +0700
+++ linux-5.10.60/net/bridge/br_sysfs_if.c	2021-08-27 14:30:58.202745192 +0700
@@ -233,6 +233,7 @@ BRPORT_ATTR_FLAG(multicast_flood, BR_MCA
 BRPORT_ATTR_FLAG(broadcast_flood, BR_BCAST_FLOOD);
 BRPORT_ATTR_FLAG(neigh_suppress, BR_NEIGH_SUPPRESS);
 BRPORT_ATTR_FLAG(isolated, BR_ISOLATED);
+BRPORT_ATTR_FLAG(block_bpdu, BR_BLOCK_BPDU);
 
 #ifdef CONFIG_BRIDGE_IGMP_SNOOPING
 static ssize_t show_multicast_router(struct net_bridge_port *p, char *buf)
@@ -286,6 +287,7 @@ static const struct brport_attribute *br
 	&brport_attr_neigh_suppress,
 	&brport_attr_isolated,
 	&brport_attr_backup_port,
+	&brport_attr_block_bpdu,
 	NULL
 };
 
diff -urpN linux-5.10.60.old/net/core/rtnetlink.c linux-5.10.60/net/core/rtnetlink.c
--- linux-5.10.60.old/net/core/rtnetlink.c	2021-08-18 13:59:19.000000000 +0700
+++ linux-5.10.60/net/core/rtnetlink.c	2021-08-27 14:30:28.998733792 +0700
@@ -4679,7 +4679,9 @@ int ndo_dflt_bridge_getlink(struct sk_bu
 	    brport_nla_put_flag(skb, flags, mask,
 				IFLA_BRPORT_MCAST_FLOOD, BR_MCAST_FLOOD) ||
 	    brport_nla_put_flag(skb, flags, mask,
-				IFLA_BRPORT_BCAST_FLOOD, BR_BCAST_FLOOD)) {
+				IFLA_BRPORT_BCAST_FLOOD, BR_BCAST_FLOOD) ||
+	    brport_nla_put_flag(skb, flags, mask,
+				IFLA_BRPORT_BLOCK_BPDU, BR_BLOCK_BPDU)) {
 		nla_nest_cancel(skb, protinfo);
 		goto nla_put_failure;
 	}
