diff -urpN linux-5.10.64.old/arch/mips/Kbuild.platforms linux-5.10.64/arch/mips/Kbuild.platforms
--- linux-5.10.64.old/arch/mips/Kbuild.platforms	2021-09-12 13:58:28.000000000 +0700
+++ linux-5.10.64/arch/mips/Kbuild.platforms	2021-09-19 22:30:30.958034765 +0700
@@ -23,6 +23,7 @@ platform-$(CONFIG_PIC32MZDA)		+= pic32/
 platform-$(CONFIG_MACH_PISTACHIO)	+= pistachio/
 platform-$(CONFIG_RALINK)		+= ralink/
 platform-$(CONFIG_MIKROTIK_RB532)	+= rb532/
+platform-$(CONFIG_RTL838X)		+= rtl838x/
 platform-$(CONFIG_SGI_IP22)		+= sgi-ip22/
 platform-$(CONFIG_SGI_IP27)		+= sgi-ip27/
 platform-$(CONFIG_SGI_IP28)		+= sgi-ip22/
diff -urpN linux-5.10.64.old/arch/mips/Kconfig linux-5.10.64/arch/mips/Kconfig
--- linux-5.10.64.old/arch/mips/Kconfig	2021-09-19 22:24:32.394257295 +0700
+++ linux-5.10.64/arch/mips/Kconfig	2021-09-19 22:33:41.977910966 +0700
@@ -627,6 +627,26 @@ config RALINK
 	select ARCH_HAS_RESET_CONTROLLER
 	select RESET_CONTROLLER
 
+config RTL838X
+	bool "Realtek based platforms"
+	select DMA_NONCOHERENT
+	select IRQ_MIPS_CPU
+	select SYS_HAS_CPU_MIPS32_R1
+	select SYS_HAS_CPU_MIPS32_R2
+	select SYS_SUPPORTS_BIG_ENDIAN
+	select SYS_SUPPORTS_HIGHMEM
+	select SYS_SUPPORTS_32BIT_KERNEL
+	select SYS_SUPPORTS_MIPS16
+	select SYS_HAS_EARLY_PRINTK
+	select BOOT_RAW
+	select PINCTRL
+	select ARCH_HAS_RESET_CONTROLLER
+	select RESET_CONTROLLER
+	select USE_OF
+	select CEVT_R4K
+	select CSRC_R4K
+
+
 config SGI_IP22
 	bool "SGI IP22 (Indy/Indigo2)"
 	select ARC_MEMORY
@@ -1034,6 +1054,31 @@ config NLM_XLP_BOARD
 
 endchoice
 
+config RTL839X
+	bool "Realtek RTL839X based platforms"
+	depends on RTL838X
+	select SYS_SUPPORTS_MULTITHREADING
+
+config RTL930X
+	bool "Realtek RTL839X based platforms"
+	depends on RTL839X
+	select MIPS_CPU_SCACHE
+	select CEVT_RTL9300
+	select SYS_SUPPORTS_MULTITHREADING
+
+config RTL931X
+	bool "Realtek RTL931X based platforms"
+	depends on RTL930X
+	select MIPS_GIC
+	select COMMON_CLK
+	select CLKSRC_MIPS_GIC
+	select SYS_SUPPORTS_VPE_LOADER
+	select SYS_SUPPORTS_SMP
+	select SYS_SUPPORTS_MIPS_CPS
+
+config CEVT_RTL9300
+	bool
+
 source "arch/mips/alchemy/Kconfig"
 source "arch/mips/ath25/Kconfig"
 source "arch/mips/ath79/Kconfig"
